providers/azure/azure.go

```diff
@@ -0,0 +1,191 @@
+			return
+		case errExpiredLeaseAzureMissing:
+			log.Printf("Cluster %s has an expired lease but doesn't exist in Azure", e.clusterName)
+			htp.Error(w, http.StatusInternalServerError, "Cluster %s has an expired lease but doesn't exist in Azure", e.clusterName)
+			return
+		default:
+			log.Printf("Unknown error %s", e.Error())
+			htp.Error(w, http.StatusInternalServerError, "Unknown error %s", e.Error())
+			return
+		}
+	}
+
+	// There is currently no way to fetch the kubeconfig from the Azure API
+	// So we must scp the file off the master node
+	kubeConfig, err := FetchKubeConfig(*availableCluster.MasterProfile.Fqdn)
+	if err != nil {
+		log.Printf("Error creating kubeconfig file for cluster %s -- %s", *availableCluster.Name, err)
+		htp.Error(w, http.StatusInternalServerError, "Error creating kubeconfig file for cluster %s -- %s", *availableCluster.Name, err)
+		return
+	}
+	newToken := uuid.NewUUID()
+	kubeConfigStr, err := k8s.MarshalAndEncodeKubeConfig(kubeConfig)
+	if err != nil {
+		log.Printf("Error marshaling & encoding kubeconfig -- %s", err)
+		htp.Error(w, http.StatusInternalServerError, "Error marshaling & encoding kubeconfig -- %s", err)
+		return
+	}
+
+	resp := api.CreateLeaseResp{
+		KubeConfigStr: kubeConfigStr,
+		IP:            *availableCluster.MasterProfile.Fqdn,
+		Token:         newToken.String(),
+		ClusterName:   *availableCluster.Name,
+	}
+
+	leaseMap.CreateLease(newToken, leases.NewLease(*availableCluster.Name, req.ExpirationTime(time.Now())))
+	if err := k8s.SaveAnnotations(services, svc, leaseMap); err != nil {
+		log.Printf("Error saving new lease to Kubernetes annotations -- %s", err)
+		htp.Error(w, http.StatusInternalServerError, "Error saving new lease to Kubernetes annotations -- %s", err)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Printf("Error encoding json -- %s", err)
+		htp.Error(w, http.StatusInternalServerError, "Error encoding json -- %s", err)
+		return
+	}
+}
+
+func getSvcsAndClusters(clusterLister ClusterLister, services k8s.ServiceGetterUpdater, k8sServiceName string) (*Map, *v1.Service, error) {
+
+	errCh := make(chan error)
+	doneCh := make(chan struct{})
+	clusterMapCh := make(chan *Map)
+	apiServiceCh := make(chan *v1.Service)
+	defer close(doneCh)
+	go func() {
+		svc, err := services.Get(k8sServiceName)
+		if err != nil {
+			select {
+			case errCh <- err:
+			case <-doneCh:
+			}
+			return
+		}
+		select {
+		case apiServiceCh <- svc:
+		case <-doneCh:
+		}
+	}()
+	go func() {
+		clusterMap, err := ParseMapFromAzure(clusterLister)
+		if err != nil {
+			select {
+			case errCh <- err:
+			case <-doneCh:
+			}
+			return
+		}
+		select {
+		case clusterMapCh <- clusterMap:
+		case <-doneCh:
+		}
+	}()
+
+	var clusterMapRet *Map
+	var apiServiceRet *v1.Service
+	for {
+		select {
+		case err := <-errCh:
+			return nil, nil, err
+		case cm := <-clusterMapCh:
+			clusterMapRet = cm
+		case svc := <-apiServiceCh:
+			apiServiceRet = svc
+		}
```

Comment:
it might be useful to have a timeout in here. [context](https://godoc.org/context) can help with that